crates/steel-parser/src/ast.rs

```diff
@@ -1011,7 +1011,22 @@ impl List {
     }
 
     pub fn make_improper(mut self) -> Self {
-        self.improper = true;
+        let Some(last) = self.args.pop() else {
+            debug_assert!(false);
+
+            self.improper = true;
+            return self;
+        };
+
+        let ExprKind::List(l) = last else {
+            self.args.push(last);
+
+            self.improper = true;
+            return self;
+        };
+
+        self.args.extend(l.args);
+        self.improper = l.improper;
```

Comment:
:information_source: This collapses e.g. `(a b . (c . d))` into `(a b c . d)`, and allows us to write stuff like `(define . (x . (1 . ())))` :upside_down_face: